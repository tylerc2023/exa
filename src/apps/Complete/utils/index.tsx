export * from './configs';
export * from './useKeyboardHeight';
export * from './userFactory';
