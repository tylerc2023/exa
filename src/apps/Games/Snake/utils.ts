export type BoardObject = Array<Array<0 | 1 | 2>>;

export const generateRandom = (size: number) =>
  Math.floor(Math.random() * size);

export const addFood = (
  board: BoardObject,
  used: {[key: string]: boolean} = {},
): boolean => {
  const x = generateRandom(board.length);
  const y = generateRandom(board.length);

  const combo = `${x}${y}`;
  if (Object.keys(used).length === board.length) {
    return false;
  }
  if (combo in used || board[x][y] !== 0) {
    used[combo] = true;
    addFood(board, used);
    return false;
  }
  board[y][x] = 2;
  return true;
};

export const addStarting = (board: BoardObject) => {
  const center = Math.floor(board.length / 2);
  board[center][center] = 1;
};

export const generateBoard = (size: number) => {
  const board: BoardObject = [];
  for (let i = 0; i < size; i++) {
    board[i] = [];
    for (let j = 0; j < size; j++) {
      board[i][j] = 0;
    }
  }

  addStarting(board);
  addFood(board);
  return board;
};

export const nextSnakePosition = () => undefined;
export const collision = (next: any) => next && false;
export const endGame = () => undefined;
export const eat = () => undefined;
