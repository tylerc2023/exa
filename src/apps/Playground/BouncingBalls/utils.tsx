import {Animated} from 'react-native';
import {State} from './BouncingBall';
import {CanvasDimensions} from './Canvas';

const getRandomNumber = (min: number, max: number): number =>
  Math.random() * (max - min) + min;

type GetSpeedProps = {
  state: State;
  canvas: CanvasDimensions;
  diameter: number;
  pulse: boolean;
};

export const getSpeed = ({state, canvas, diameter, pulse}: GetSpeedProps) => {
  const dMax = state.diameter >= diameter * 1.1;
  const dMin = state.diameter <= diameter * 0.9;
  // TODO: recenter so pulsing happens from middle
  const growing = dMax || dMin ? !state.growing : state.growing;
  const d = state.diameter + (!pulse ? 0 : growing ? 1 : -1);
  const xMin = canvas.x;
  const xMax = canvas.x + canvas.width;
  const yMin = canvas.y;
  const yMax = canvas.y + canvas.height;
  const xLarger = state.x + d >= xMax;
  const xSmaller = state.x <= xMin;
  const yLarger = state.y + d >= yMax;
  const ySmaller = state.y <= yMin;
  // TODO: add collision
  const xSpeed = xLarger || xSmaller ? -state.xSpeed : state.xSpeed;
  const ySpeed = yLarger || ySmaller ? -state.ySpeed : state.ySpeed;
  const x = xLarger ? xMax - d : xSmaller ? xMin : state.x + xSpeed;
  const y = yLarger ? yMax - d : ySmaller ? yMin : state.y + ySpeed;
  const dx = x + xSpeed;
  const dy = y + ySpeed;
  return {xSpeed, ySpeed, x: dx, y: dy, d, growing};
};

type GetInitialStateProps = {
  canvas: CanvasDimensions;
  diameter: number;
  speed: number;
};

export const getInitialState = ({
  canvas,
  diameter,
  speed,
}: GetInitialStateProps): State => {
  const starting = {
    x: getRandomNumber(canvas.x, canvas.x + canvas.width - diameter),
    y: getRandomNumber(canvas.y, canvas.y + canvas.height - diameter),
  };
  const growing = !!Math.floor(Math.random());
  return {
    ...starting,
    position: new Animated.ValueXY(starting),
    xSpeed: getRandomNumber(-speed, speed),
    ySpeed: getRandomNumber(-speed, speed),
    diameter,
    growing,
  };
};
