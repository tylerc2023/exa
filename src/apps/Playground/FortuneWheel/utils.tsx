type GetRandomRange = {
  min: number;
  max: number;
};

const getRandomRange = ({min, max}: GetRandomRange) => {
  return Math.floor(Math.random() * (max - min) + min);
};

type GetNewLocation = {
  minSpin: number;
  maxSpin: number;
  numOfSegments: number;
  location: number;
};

export const getNewLocation = ({
  minSpin,
  maxSpin,
  numOfSegments,
  location,
}: GetNewLocation) => {
  const spins = getRandomRange({min: minSpin, max: maxSpin}) * 360;
  const winner = Math.floor(Math.random() * numOfSegments);
  const winnerSection = Math.floor(winner * (360 / numOfSegments));
  const newLocation = location + spins + winnerSection;
  return newLocation;
};

type GetWinnerIndex = {
  location: number;
  numOfSegments: number;
  angleOfSegment: number;
};

export const getWinnerIndex = ({
  location,
  numOfSegments,
  angleOfSegment,
}: GetWinnerIndex) => {
  const deg = Math.abs(Math.round(location % 360));
  const index =
    (numOfSegments - Math.floor(deg / angleOfSegment)) % numOfSegments;
  return index;
};
