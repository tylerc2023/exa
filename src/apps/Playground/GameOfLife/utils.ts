const operations = [
  [0, 1],
  [0, -1],
  [1, -1],
  [-1, 1],
  [1, 1],
  [-1, -1],
  [1, 0],
  [-1, 0],
];

type BoardItem = 0 | 1;

export type Board = Array<Array<BoardItem>>;

export const generateBoard = (size: number, random = 0): Board => {
  const board: Board = [];
  for (let i = 0; i < size; i++) {
    board[i] = [];
    for (let j = 0; j < size; j++) board[i][j] = Math.random() > random ? 0 : 1;
  }
  return board;
};

export const determineBoardItem = (
  board: Board,
  i: number,
  j: number,
): BoardItem => {
  let neighbors = 0;
  const item = board[i][j];
  const length = board.length;
  for (let k = 0; k < operations.length; k++) {
    const [x, y] = operations[k];
    const newI = i + x;
    const newK = j + y;
    if (newI >= 0 && newI < length && newK >= 0 && newK < length)
      neighbors += board[newI][newK];
    if (neighbors > 3) break;
  }

  return neighbors < 2 || neighbors > 3
    ? 0
    : item === 0 && neighbors === 3
    ? 1
    : item;
};
