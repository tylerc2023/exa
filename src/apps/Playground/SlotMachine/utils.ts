import {Combinations, Reel, Reels} from './config';

const getWinsPerCombination = (
  combinations: Combinations,
  reels: Reels,
): Combinations => {
  const output = [...combinations];
  for (let i = 0; i < combinations.length; i++) {
    let freq = 1;
    output[i] = [...combinations[i]];
    for (let j = 0; j < reels.length; j++) {
      const symbol = output[i][j];
      const reel = reels[j];
      const sum = getSumOfReel(reel);
      freq *=
        symbol !== 'any' && typeof symbol !== 'number' ? reel[symbol] : sum;
    }
    output[i][3] = freq;
  }
  return output;
};

const getSumOfReel = (reel: Reel): number => {
  return Object.values(reel).reduce((sum, item) => sum + item, 0);
};

const getSumOfReels = (reels: Reels): number => {
  return reels.reduce((total, reel) => total * getSumOfReel(reel), 1);
};

const getReturnsPerCombination = (
  combinations: Combinations,
  winsPerCombination: Combinations,
): Combinations => {
  return winsPerCombination.reduce(
    (output: Combinations, combination, index) => {
      const payout = combinations[index][3];
      output[index][3] = combination[3] * payout;
      return output;
    },
    [...winsPerCombination],
  );
};

const getSumOfCombinations = (combinations: Combinations): number => {
  return combinations.reduce((total, combination) => total + combination[3], 0);
};

export const getWinPercentage = (
  combinations: Combinations,
  reels: Reels,
): number => {
  const sumOfReels = getSumOfReels(reels);
  const winsPerCombination = getWinsPerCombination(combinations, reels);
  const sumOfWins = getSumOfCombinations(winsPerCombination);
  return sumOfWins / sumOfReels;
};

export const getReturnPercentage = (
  combinations: Combinations,
  reels: Reels,
): number => {
  const sumOfReels = getSumOfReels(reels);
  const winsPerCombination = getWinsPerCombination(combinations, reels);
  const returnsPerCombination = getReturnsPerCombination(
    combinations,
    winsPerCombination,
  );
  const sumOfReturns = getSumOfCombinations(returnsPerCombination);
  return sumOfReturns / sumOfReels;
};
