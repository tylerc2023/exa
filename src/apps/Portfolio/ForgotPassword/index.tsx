import React, {memo, useCallback, useState} from 'react';
import {StyleSheet, View} from 'react-native';
import {Button, Screen, TextInput} from '../../../components';
import {useNav} from '../../../hooks';

const styles = StyleSheet.create({
  row: {flexDirection: 'row', justifyContent: 'space-around'},
});

export const ForgotPassword = memo(function PortfolioForgotPassword() {
  const [email, setEmail] = useState('');
  const nav = useNav();
  const handleEmail = useCallback((val: string) => setEmail(val), []);
  const navLogin = useCallback(nav('login'), [nav]);
  return (
    <Screen gutter onLeftPress={navLogin} title="Forgot Password">
      <TextInput
        onChangeText={handleEmail}
        placeholder="example@gmail.com"
        title="email"
        value={email}
      />
      <View style={styles.row}>
        <Button onPress={navLogin} title="Send email" />
      </View>
    </Screen>
  );
});
