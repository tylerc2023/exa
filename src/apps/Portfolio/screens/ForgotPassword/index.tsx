import React, {memo, useCallback, useState} from 'react';
import {StyleSheet, View} from 'react-native';
import {Button, Screen, TextInput} from '../../../../components';
import {useNav} from '../../../../hooks';

const styles = StyleSheet.create({
  row: {flexDirection: 'row', justifyContent: 'space-around'},
});

export default memo(function PortfolioForgotPassword() {
  const [email, setEmail] = useState('');
  const nav = useNav();
  const handleEmail = useCallback((val: string) => setEmail(val), []);
  return (
    <Screen
      gutter
      onLeftPress={nav.to('portfolioLogin')}
      title="Forgot Password">
      <TextInput
        onChangeText={handleEmail}
        placeholder="example@gmail.com"
        title="email"
        value={email}
      />
      <View style={styles.row}>
        <Button onPress={nav.to('portfolioLogin')} title="Send email" />
      </View>
    </Screen>
  );
});
