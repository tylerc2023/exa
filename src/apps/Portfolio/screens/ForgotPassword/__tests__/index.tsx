import React from "react";
import { mockRenderer } from "../../../../../utils/Mock";
import { ForgotPassword } from "../../ForgotPassword";

it("renders correctly", () => {
  const dom = mockRenderer(<ForgotPassword />).toJSON();
  expect(dom).toMatchSnapshot();
});
