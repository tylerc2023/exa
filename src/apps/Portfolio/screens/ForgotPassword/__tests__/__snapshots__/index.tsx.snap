// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "hsl(0, 0%, 100%)",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "hsl(210, 10%, 23%)",
        "borderBottomWidth": 0,
        "flexDirection": "row",
        "height": 72,
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 0.2,
          "opacity": 1,
          "padding": 12,
        }
      }
    >
      <View>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "hsl(210, 10%, 23%)",
                "fontSize": 24,
              },
              Array [
                undefined,
                undefined,
              ],
              Object {
                "fontFamily": "Material Design Icons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
    <Text
      style={
        Object {
          "color": "hsl(0, 0%, 0%)",
          "fontSize": 30,
          "fontWeight": "100",
          "letterSpacing": -0.5,
          "opacity": 1,
        }
      }
    >
      Forgot Password
    </Text>
    <View
      accessible={true}
      focusable={false}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 0.2,
          "opacity": 1,
          "padding": 12,
        }
      }
    />
  </View>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={
      Array [
        Object {
          "backgroundColor": "hsl(0, 0%, 100%)",
          "flex": 1,
        },
        undefined,
      ]
    }
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    keyboardSpace={0}
    onScroll={[Function]}
    resetKeyboardSpace={[Function]}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <View
        style={
          Array [
            undefined,
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "flex-start",
                "backgroundColor": "transparent",
                "borderColor": "transparent",
                "borderRadius": 4,
                "borderWidth": 1,
                "flexDirection": "row",
                "height": 20,
                "justifyContent": "flex-start",
                "marginVertical": 4,
                "opacity": 1,
                "paddingHorizontal": 0,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "hsl(0, 0%, 0%)",
                  "fontSize": 14,
                  "fontWeight": "300",
                  "letterSpacing": 0.25,
                  "opacity": 1,
                  "textAlign": "center",
                }
              }
            >
              email
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <TextInput
            allowFontScaling={true}
            blurOnSubmit={true}
            disableFullscreenUI={true}
            editable={true}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="example@gmail.com"
            placeholderTextColor="hsl(208, 7%, 46%)"
            rejectResponderTermination={true}
            selectionColor="hsl(211, 100%, 50%)"
            style={
              Array [
                Object {
                  "backgroundColor": "hsl(0, 0%, 100%)",
                  "borderColor": "hsl(208, 7%, 46%)",
                  "borderRadius": 4,
                  "borderWidth": 2,
                  "color": "hsl(0, 0%, 0%)",
                  "padding": 8,
                  "paddingRight": 32,
                  "width": "100%",
                },
                "",
                false,
                Object {
                  "fontSize": 14,
                  "fontWeight": "300",
                  "letterSpacing": 0.25,
                },
                undefined,
              ]
            }
            underlineColorAndroid="transparent"
            value=""
          />
        </View>
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "flex-start",
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 20,
              "justifyContent": "flex-start",
              "marginVertical": 4,
              "opacity": 1,
              "paddingHorizontal": 0,
            }
          }
        >
          <View
            style={
              Array [
                "",
                Array [
                  Object {
                    "color": "hsl(211, 100%, 50%)",
                  },
                  undefined,
                  undefined,
                  Object {
                    "color": "hsl(354, 70%, 54%)",
                  },
                  undefined,
                  undefined,
                  undefined,
                ],
              ]
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "hsl(210, 10%, 23%)",
                    "fontSize": 16,
                  },
                  Array [
                    Object {
                      "opacity": 0,
                    },
                    undefined,
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
          <Text
            style={
              Object {
                "color": "hsl(354, 70%, 54%)",
                "fontSize": 14,
                "fontWeight": "300",
                "letterSpacing": 0.25,
                "opacity": 1,
                "textAlign": "center",
              }
            }
          >
            
          </Text>
        </View>
      </View>
      <View
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-around",
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          isTVSelectable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderRadius": 4,
              "borderWidth": 1,
              "flexDirection": "row",
              "height": 36,
              "justifyContent": "center",
              "opacity": 1,
              "paddingHorizontal": 16,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "hsl(211, 100%, 50%)",
                "fontSize": 14,
                "fontWeight": "600",
                "letterSpacing": 0.75,
                "opacity": 1,
                "textAlign": "center",
              }
            }
          >
            SEND EMAIL
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
